src/gmmbin/gmm-init-biphone.cc

```diff
@@ -0,0 +1,245 @@
+          for (size_t ik = 0; ik < ipset.size(); ik++) {
+            level2_map[ipset[ik]] = new TableEventMap(kPdfClass, level3_map);
+          }
+        }
+        level1_map[pset[k]] = new TableEventMap(0, level2_map); //lvl2_eventmap;
+        if (k != pset.size() - 1)
+          current_pdfid = base_pdfid;
+      } /////////   k
+
+    }
+  }
+
+  return new TableEventMap(1, level1_map);
+}
+
+
+ContextDependency*
+BiphoneContextDependencyFull(const std::vector<std::vector<int32> > phone_sets,
+                             const std::vector<int32> phone2num_pdf_classes) {
+  std::vector<bool> share_roots(phone_sets.size(), false);  // don't share roots
+  // N is context size, P = position of central phone (must be 0).
+  int32 P = 1, N = 2;
+  EventMap *pdf_map = GetFullBiphoneStubMap(phone_sets,
+                                            phone2num_pdf_classes, share_roots);
+  return new ContextDependency(N, P, pdf_map);
+}
+
+
+} // end namespace kaldi
+
+int main(int argc, char *argv[]) {
+  try {
+    using namespace kaldi;
+    using kaldi::int32;
+
+    const char *usage =
+        "Initialize a biphone context-dependency tree with all the\n"
+        "leaves (i.e. a full tree). Intended for end-to-end tree-free models.\n"
+        "Usage:  gmm-init-biphone <topology-in> <dim> <model-out> <tree-out> \n"
+        "e.g.: \n"
+        " gmm-init-biphone topo 39 bi.mdl bi.tree\n";
+
+    bool binary = true;
+    std::string shared_phones_rxfilename;
+    ParseOptions po(usage);
+    po.Register("binary", &binary, "Write output in binary mode");
+    po.Register("shared-phones", &shared_phones_rxfilename,
+                "rxfilename containing, on each line, a list of phones whose pdfs should be shared.");
+    po.Read(argc, argv);
+
+    if (po.NumArgs() != 4) {
+      po.PrintUsage();
+      exit(1);
+    }
+
+
+    std::string topo_filename = po.GetArg(1);
+    int dim = atoi(po.GetArg(2).c_str());
+    KALDI_ASSERT(dim> 0 && dim < 10000);
+    std::string model_filename = po.GetArg(3);
+    std::string tree_filename = po.GetArg(4);
+
+    Vector<BaseFloat> glob_inv_var(dim);
+    glob_inv_var.Set(1.0);
+    Vector<BaseFloat> glob_mean(dim);
+    glob_mean.Set(1.0);
+
+    HmmTopology topo;
+    bool binary_in;
+    Input ki(topo_filename, &binary_in);
+    topo.Read(ki.Stream(), binary_in);
+
+    const std::vector<int32> &phones = topo.GetPhones();
+
+    std::vector<int32> phone2num_pdf_classes (1 + phones.back());
+    for (size_t i = 0; i < phones.size(); i++) {
+      phone2num_pdf_classes[phones[i]] = topo.NumPdfClasses(phones[i]);
+      // for now we only support 1 or 2 pdf per phone
+      KALDI_ASSERT(phone2num_pdf_classes[phones[i]] == 1 ||
+                   phone2num_pdf_classes[phones[i]] == 2);
+    }
+
+    // Now the tree:
+    ContextDependency *ctx_dep = NULL;
+    std::vector<std::vector<int32> > shared_phones;
+    if (shared_phones_rxfilename == "") {
+      for (size_t i = 0; i < phones.size(); i++)
```

Comment:
Can you add braces to this for loop?

Reply:
It's a single-line statement, why are suggesting "braces"?

Reply:
Just a common best practice to always use curly braces, especially since your single-line statement (in my opinion) is rather long.

Reply:
Actually the Google style guide doesn't mandate them for one-line statements, and in cases like this I prefer not to have them as they make the code longer.